### ISSUE TITLE
Docs falsely suggest `exec()` for async/await

### ISSUE THREAD

**Issue by Moeriki** [Automattic/mongoose#8110]
**Do you want to request a *feature* or report a *bug*?**

Docs.

**What is the current docs?**

https://mongoosejs.com/docs/promises.html#built-in-promises

> This means that you can do things like MyModel.findOne({}).then() and await MyModel.findOne({}).exec() if you're using async/await.

**What is the expected docs?**

`await Model.findOne()` works just fine.

> This means that you can do things like `MyModel.findOne({}).then()` and `await MyModel.findOne({})` if you're using async/await.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

Mongoose 5.5.14
Node.js 10.16.2

**Comment by vkarpov15** [comment]
Both approaches work. We'll add a note to the docs.

**Comment by Moeriki** [comment]
I'm not sure why you would reference both, since the `.exec()` is just more code with no benefit. Or am I missing something?

I found out by means of a colleague who unknowingly wrote this since the docs suggested it. As far as I can tell `.exec()` is only useful for invoking a callback-style response.

I could PR docs.

**Comment by jloveridge** [comment]
@Moeriki The difference is that the return type of `Model.findOne` is a `Query`. They just happen to have added a `then` to it so it is "thennable" but it is not a true promise. The return from `exec` is actually a promise if you don't supply a callback to the `exec`. While you might not care about this distinction some of us do.

**Comment by Moeriki** [comment]
I see. Thanks for the feedback.

I take back my words. Since users may have certain expectations from Promises I'd keep the `.exec()` as the most prominent documentation. We could still mention the "thennable" object in smaller print.

**Comment by vkarpov15** [comment]
You're right that queries are not promises: https://mongoosejs.com/docs/queries.html#queries-are-not-promises . The most meaningful difference is that if you `await` on the same query object twice, you will execute the query twice, whereas `await` on the same promise twice only executes the query once. We did also run into an issue where node's async_hooks don't work properly if you `await` on a custom thennable, but that's a node issue that mongoose can't work around.

TLDR; there are reasons to prefer using `exec()`, but for most apps it doesn't matter which one.

@jloveridge I'm curious to hear why you prefer a true promise over a thennable. Is it one of the reasons I listed or something else?

**Comment by Enado95** [comment]
So @vkarpov15  would my approach work :

`async(companyID) => { await User.find({company:companyID}).then(data =>{ return data }).catch(err => { return err.message }) } `

**Comment by vkarpov15** [comment]
@Enado95 I'm not sure what you mean by 'work'. The code as written definitely needs some work - its clumsy and suppresses errors. Why do you do `.then(data => data)`? That does nothing besides create an unnecessary promise. Also, why the `.catch()`? That means that you ignore all errors.

**Comment by Enado95** [comment]
> @Enado95 I'm not sure what you mean by 'work'. The code as written definitely needs some work - its clumsy and suppresses errors. Why do you do `.then(data => data)`? That does nothing besides create an unnecessary promise. Also, why the `.catch()`? That means that you ignore all errors.

@vkarpov15 When say it works. I mean it returns a response and if there's mongo error it returns the a 500 response as well; here's my exact implementation below for deleting a record. I do the same for updating and save():

```javascript
exports.deleteCompany = async (req, res) => {
  const company = await Company.findByIdAndDelete(
    mongoose.Types.ObjectId(req.params.id)
  )
    .then(data => {
      if (!data)
        return res.status(404).send('Company with the given id was not found')

      res.send('Company successfully deleted')
    })
    .catch(err => {
      return res.status(500).send({
        message: err.message
      })
    })
}
```
Since the `.then(data => data)` does nothing, how would truly make this method async and catch errors?

**Comment by vkarpov15** [comment]
This implementation for deleting a company should work fine. Not sure what you mean by "truly make this method async".

**Comment by Enado95** [comment]
So the same implementation for find or findOneAndUpdate would work similarly? 
Also, since its and async function, is the `.exec()` required?

**Comment by vkarpov15** [comment]
find() and findOneAndUpdate() both return mongoose queries, so they should be fine as far as `await` is concerned. The `exec()` isn't necessary.

**Comment by Enado95** [comment]
Okay noted.

**Comment by DevBrent** [comment]
> So @vkarpov15 i use this approach :
> 
> `async(companyID) => { await User.find({company:companyID}).then(data =>{ return data }).catch(err => { return err.message }) } `
> 
> It works but is it faulty? I do the same for updating, saving and deleting documents
> Or is it that i should do mongoose.Promose = require('bluebird') to make this "work" as intended

A string is not an error. I would avoid throwing strings by any means necessary.

https://web.archive.org/web/20111226063046/http://www.devthought.com/2011/12/22/a-string-is-not-an-error/
